### ISSUE TITLE
Auto update and semantic versioning?

### ISSUE THREAD

**Issue by ochedru** [containrrr/watchtower#111]
Our docker images are versioned using [semantic versioning](http://semver.org/). Image tags typically end with `major.minor.patch` numbers.
We would like to stay in control for major or minor upgrades, because they are likely to break things. So, I was wondering if there is a way to configure watchtower to update a container only if a new patch is available?
Thank you!

**Comment by stffabi** [comment]
The best would be if you follow the guidelines of the docker official images for [tags and aliases](https://github.com/docker-library/official-images#tags-and-aliases). Following that you could just use e.g. the tag 2.3 and you automatically stay on the newest patch version of 2.3. `2.3` is essentially an alias for `2.3.0`, `2.3.1`, `2.3.2`,... as soon as they get released.

**Comment by ochedru** [comment]
Indeed, this is the best solution! Thank you.

**Comment by stffabi** [comment]
You're welcome.

**Comment by benridley** [comment]
For anyone stumbling across this issue in the future, I looked into adding the idea of a 'tag pattern' to Watchtower, so that if your containers had a label with a pattern, Watchtower would only update using images that matched the pattern. For example, you could restrict updates to '8\..*' to pin yourself to 8.x.x in the example of semver.

As it turns out, this is not so easy. Watchtower works by simply pulling the same image/tag, and comparing the content. Providing your image follows the guidelines above, this works very well. For my idea, you would need to actually pull all the tags of the image to perform the comparison, and as far as I know there's no straighforward way to do this using public image repositories. Even if there was, some repositories have enormous amounts of tags (some of them push every git commit to the repository, going back years), so this would perform pretty badly and probably consume a lot of resources while matching tags.

TL;DR: The guidelines above are a much better way of approaching this problem. If the Docker image you are interested in does not use the guidelines, a lot of open source software is repackaged by other parties like Bitnami which use this standard, so you can see if they offer your image as an alternative.

**Comment by jlssmt** [comment]
@stffabi @benridley
The only correct solution would be imho to pin a package to major/patch like mentioned in answer 1, but additionally get watchtower to watch for new major/patch and notify me so I can upgrade big changes manually in the near future.
Any chances to get this into watchtower? This would be game changer for stability + security.

**Comment by piksel** [comment]
@jlssmt Checking for new tags would be wholly new code and should probably be done in a separate project. It would be a matter of making explicit requests to the repositories to fetch the metadata. Maybe it could be added as a separate tool to the repo if someone is willing to write it though.

**Comment by JanJK2011** [comment]
I scripted [here](https://github.com/jansppenrade2/Docker-Container-Updater) something for this purpose. Tested on CentOS Stream 9 and Qnap QTS.
